File: templates/python/analysis_template.py
"""Paper Analysis 4SS - Python template."""

from __future__ import annotations

import os
import re
from pathlib import Path

import numpy as np
import pandas as pd
import statsmodels.formula.api as smf


OUTPUT_ROOT = Path(os.environ.get("OUTPUT_ROOT", "output/paper-analysis"))
for sub in [
    "data",
    "scripts",
    "tables",
    "figures",
    "reports",
    "qual/codebooks",
    "qual/coded-data",
    "qual/anonymized",
    "qual/reliability",
]:
    (OUTPUT_ROOT / sub).mkdir(parents=True, exist_ok=True)


def load_data(path: str | Path) -> pd.DataFrame:
    path = Path(path)
    ext = path.suffix.lower().lstrip(".")
    if ext == "csv":
        return pd.read_csv(path)
    if ext in {"xlsx", "xls"}:
        return pd.read_excel(path)
    if ext == "dta":
        return pd.read_stata(path)
    if ext == "parquet":
        return pd.read_parquet(path)
    raise ValueError(f"Unsupported file type: {ext}")


def clean_data(df: pd.DataFrame) -> pd.DataFrame:
    out = df.copy()
    for col in out.select_dtypes(include="object"):
        out[col] = out[col].astype(str).str.strip().replace({"": np.nan, "nan": np.nan})
    return out.drop_duplicates()


def sample_flow(df: pd.DataFrame, y: str, x: str, controls: list[str]) -> pd.DataFrame:
    vars_needed = [y, x, *controls]
    rows = []
    current = df.copy()
    rows.append({"step": "raw", "n": len(current)})
    for var in vars_needed:
        current = current[current[var].notna()]
        rows.append({"step": f"nonmissing_{var}", "n": len(current)})
    flow = pd.DataFrame(rows)
    flow.to_csv(OUTPUT_ROOT / "tables/sample-flow.csv", index=False)
    return flow


def describe_data(df: pd.DataFrame, vars_: list[str]) -> pd.DataFrame:
    table = df[vars_].describe(include="all").T
    numeric_cols = table.select_dtypes(include=[np.number]).columns
    table[numeric_cols] = table[numeric_cols].round(3)
    table.to_csv(OUTPUT_ROOT / "tables/table1-descriptives.csv")
    return table


def run_ols(df: pd.DataFrame, y: str, x: str, controls: list[str], cluster: str | None = None):
    rhs = " + ".join([x, *controls])
    model = smf.ols(f"{y} ~ {rhs}", data=df).fit()
    if cluster:
        model = model.get_robustcov_results(cov_type="cluster", groups=df[cluster])
    else:
        model = model.get_robustcov_results(cov_type="HC3")
    return model


def run_logit_ame(df: pd.DataFrame, y: str, x: str, controls: list[str]):
    rhs = " + ".join([x, *controls])
    model = smf.logit(f"{y} ~ {rhs}", data=df).fit(disp=False)
    ame = model.get_margeff(at="overall", method="dydx")
    return model, ame


def export_regression_text(model, name: str = "table2-main-regression") -> None:
    with open(OUTPUT_ROOT / f"tables/{name}.txt", "w", encoding="utf-8") as f:
        f.write(model.summary().as_text())
        f.write("\n\nNote: use the companion CSV for publication tables with t/z statistics below coefficients.\n")


def export_regression_table(model, name: str = "table2-main-regression") -> pd.DataFrame:
    """Export a publication-style table with t/z statistics below coefficients."""
    params = pd.Series(model.params, index=model.model.exog_names, name="coef")
    stat_values = pd.Series(model.tvalues, index=model.model.exog_names, name="statistic")
    pvalues = pd.Series(model.pvalues, index=model.model.exog_names, name="p")

    rows = []
    for term in params.index:
        coef = params.loc[term]
        stat = stat_values.loc[term]
        p = pvalues.loc[term]
        stars = "***" if p < 0.001 else "**" if p < 0.01 else "*" if p < 0.05 else ""
        rows.append({"term": term, "row": "coef", "estimate": f"{coef:.3f}{stars}"})
        rows.append({"term": term, "row": "t_or_z", "estimate": f"({stat:.3f})"})

    table = pd.DataFrame(rows)
    table.to_csv(OUTPUT_ROOT / f"tables/{name}.csv", index=False)
    return table


def anonymize_text(text: str) -> str:
    text = re.sub(r"[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\.[A-Za-z]{2,}", "[EMAIL REMOVED]", text)
    text = re.sub(r"(\+?\d[\d\-\s]{7,}\d)", "[PHONE REMOVED]", text)
    text = re.sub(r"\b\d{3}-\d{2}-\d{4}\b", "[ID REMOVED]", text)
    text = re.sub(r"\b1[3-9]\d{9}\b", "[PHONE REMOVED]", text)
    text = re.sub(r"\b\d{17}[\dXx]\b", "[ID REMOVED]", text)
    text = re.sub(r"\b\d{6}(19|20)\d{2}(0[1-9]|1[0-2])([0-2]\d|3[01])\d{3}[\dXx]\b", "[ID REMOVED]", text)
    text = re.sub(r"(微信|WeChat|QQ)[:：]?\s*[A-Za-z0-9_\-]{5,}", r"\1: [ACCOUNT REMOVED]", text)
    text = re.sub(r"\d{4}[年/-]\d{1,2}[月/-]\d{1,2}日?", "[DATE REMOVED]", text)
    return text


def segment_text_file(path: str | Path, document_id: str | None = None) -> pd.DataFrame:
    path = Path(path)
    document_id = document_id or path.stem
    text = anonymize_text(path.read_text(encoding="utf-8"))
    paragraphs = [p.strip() for p in re.split(r"\n\s*\n", text) if p.strip()]
    df = pd.DataFrame(
        {
            "document_id": document_id,
            "segment_id": [f"{document_id}-{i+1:04d}" for i in range(len(paragraphs))],
            "text": paragraphs,
        }
    )
    df.to_csv(OUTPUT_ROOT / f"qual/coded-data/segments-{document_id}.csv", index=False)
    return df


def code_frequency(coded_data: pd.DataFrame) -> pd.DataFrame:
    freq = coded_data["code"].value_counts().rename_axis("code").reset_index(name="n")
    freq["pct"] = freq["n"] / freq["n"].sum() * 100
    freq.to_csv(OUTPUT_ROOT / "tables/qual-code-frequency.csv", index=False)
    return freq


if __name__ == "__main__":
    print("Edit this template with project-specific paths and variables before running.")

File: templates/r/analysis-template.R
# Paper Analysis 4SS - R template

suppressPackageStartupMessages({
  library(tidyverse)
  library(haven)
  library(readxl)
  library(modelsummary)
  library(fixest)
  library(marginaleffects)
  library(ggplot2)
})

output_root <- Sys.getenv("OUTPUT_ROOT", "output/paper-analysis")
dir.create(output_root, recursive = TRUE, showWarnings = FALSE)
for (d in c("data", "scripts", "tables", "figures", "reports", "qual/codebooks", "qual/coded-data", "qual/reliability")) {
  dir.create(file.path(output_root, d), recursive = TRUE, showWarnings = FALSE)
}

load_data <- function(path) {
  ext <- tolower(tools::file_ext(path))
  switch(ext,
    csv = readr::read_csv(path, show_col_types = FALSE),
    xlsx = readxl::read_excel(path),
    xls = readxl::read_excel(path),
    dta = haven::read_dta(path),
    sav = haven::read_sav(path),
    rds = readRDS(path),
    parquet = arrow::read_parquet(path),
    stop("Unsupported file type: ", ext)
  )
}

clean_data <- function(df) {
  df %>%
    mutate(across(where(is.character), ~na_if(str_squish(.x), ""))) %>%
    distinct()
}

make_sample <- function(df, y, x, controls = character()) {
  vars <- c(y, x, controls)
  df %>%
    mutate(analysis_sample = if_all(all_of(vars), ~ !is.na(.x))) %>%
    filter(analysis_sample)
}

make_table1 <- function(df, vars) {
  datasummary_skim(
    df[, vars, drop = FALSE],
    fmt = 3,
    output = file.path(output_root, "tables/table1-descriptives.html")
  )
}

run_ols <- function(df, y, x, controls = character(), fe = NULL, cluster = NULL) {
  rhs <- paste(c(x, controls), collapse = " + ")
  if (!is.null(fe) && length(fe) > 0) {
    fml <- as.formula(paste(y, "~", rhs, "|", paste(fe, collapse = " + ")))
  } else {
    fml <- as.formula(paste(y, "~", rhs))
  }
  vc <- if (!is.null(cluster)) as.formula(paste("~", cluster)) else "HC3"
  feols(fml, data = df, vcov = vc)
}

run_logit_ame <- function(df, y, x, controls = character()) {
  fml <- as.formula(paste(y, "~", paste(c(x, controls), collapse = " + ")))
  mod <- glm(fml, data = df, family = binomial())
  ame <- avg_slopes(mod, variables = x)
  list(model = mod, ame = ame)
}

export_models <- function(models) {
  table_args <- list(
    statistic = "statistic",
    fmt = 3,
    stars = c("*" = .05, "**" = .01, "***" = .001),
    notes = "括号内为 t/z 统计值；标准误类型和聚类层级见正文或表注。"
  )
  do.call(modelsummary, c(list(models = models, output = file.path(output_root, "tables/table2-main-regression.html")), table_args))
  do.call(modelsummary, c(list(models = models, output = file.path(output_root, "tables/table2-main-regression.tex")), table_args))
  do.call(modelsummary, c(list(models = models, output = file.path(output_root, "tables/table2-main-regression.docx")), table_args))
  do.call(modelsummary, c(list(models = models, output = file.path(output_root, "tables/table2-main-regression.csv")), table_args))
}

plot_coef <- function(model) {
  p <- modelplot(model) +
    geom_vline(xintercept = 0, linetype = "dashed") +
    theme_minimal(base_size = 12) +
    labs(x = "Estimate", y = NULL)
  ggsave(file.path(output_root, "figures/coefplot-main.pdf"), p, width = 7, height = 5)
  ggsave(file.path(output_root, "figures/coefplot-main.png"), p, width = 7, height = 5, dpi = 300)
}

code_reliability <- function(coder_a, coder_b) {
  if (!requireNamespace("irr", quietly = TRUE)) {
    stop("Install irr to compute Cohen's kappa: install.packages('irr')")
  }
  irr::kappa2(data.frame(coder_a, coder_b))
}

# Example usage:
# df <- load_data("data/raw/data.csv") |> clean_data()
# df_a <- make_sample(df, y = "outcome", x = "treatment", controls = c("age", "education"))
# make_table1(df_a, c("outcome", "treatment", "age", "education"))
# m1 <- run_ols(df_a, "outcome", "treatment", c("age", "education"), fe = c("year", "region"), cluster = "region")
# export_models(list("Main" = m1))
# plot_coef(m1)
